### PR TITLE
Pathfinder

### PR DESCRIPTION
**Pull request by HilelAmiel**
this branch adds jaci's motion profiling libs

### CODE REVIEW

File: MotionProfiling_Project/src/main/java/frc/robot/SettingsForPID.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
+
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/SettingsForPID.java]: תוריד את הרווח!

File: MotionProfiling_Project/src/main/java/frc/robot/SettingsForPID.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+public class SettingsForPID {
+
+    double KP, KI, KD, KV, KA, tolerance, waitTime;
+    /**
+    * @param KP the the Proportional coefficient of the PID loop in this command.
+    * @param KI the Integral coefficient of the PID loop in this command.
+    * @param KD the Differential coefficient of the PID loop in this command.
+    * @param KV the velocity at which to go.
+    * @param KA the acceleration of the robot
+    * @param tolerance  the error tolerance of this command.
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/SettingsForPID.java]: change the description, kv ka, רווח בטולורנס

File: MotionProfiling_Project/src/main/java/frc/robot/SettingsForPID.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        this.KP = KP;
+        this.KI = KI;
+        this.KD = KD;
+        this.tolerance = tolerance;
+        this.waitTime = waitTime;
+
+    }
+
+    public SettingsForPID(double KP, double KD, double KV, double KA) {
+        this.KP = KP;
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/SettingsForPID.java]: change to PID Settings pls

File: MotionProfiling_Project/src/main/java/frc/robot/RobotConstants.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    public static final double WHEEL_DIAMETER = 0.1524;
+    public static final double WHEEL_BASE_WIDTH = 0.6;
+    public static final double TIMEFRAME = 0.02;
+
+    public static final double MAX_VELOCITY = 3;
+    public static final double MAX_ACCELERATION = 2;
+    public static final double MAX_JERK = 80;
+
+    public static final SettingsForPID MotionProfilingPIDSettings = new SettingsForPID(0, 0, 80, 0);
+
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/RobotConstants.java]: big latters

File: MotionProfiling_Project/src/main/java/frc/robot/RobotConstants.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    public static final double kP_TURN = 0.8;
+
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/RobotConstants.java]: motion profiling kp turn

File: MotionProfiling_Project/src/main/java/frc/robot/RobotConstants.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    public static final double RIGHT_TICKS_DIVIDER = 575.0;
+    public static final double LEFT_TICKS_DIVIDER = 771.5;
+    public static final int TICKS_PER_REV = 50;
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/RobotConstants.java]: ticks per meter right / left
revolution

File: MotionProfiling_Project/src/main/java/frc/robot/RobotConstants.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
+package frc.robot;
+
+/**
+ * all the constants on the code are stored here.
+ */
+public class RobotConstants {
+
+    
+    public static final double WHEEL_DIAMETER = 0.1524;
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/RobotConstants.java:14]: too many spaces

File: MotionProfiling_Project/src/main/java/frc/robot/PathCreater.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+        generate_AllPaths();
+    }
+
+    /** takes all the paths we created and generates them */
+    public void generate_AllPaths() {
+        for (int i = 0; i < this.paths.size(); i++) {
+            this.trajectories[i] = Pathfinder.generate(this.paths.get(i), this.config);
+
+        }
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/PathCreater.java]: space remove

File: MotionProfiling_Project/src/main/java/frc/robot/PathCreater.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    /**
+     * we split the tank drive for bpth sides of the robot and get the trajectory of
+     * each one, first we get the right then the left trajectory
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/PathCreater.java]: both, "the func gets the path number to split and return array of trajector, index 0 for the right side and 1 for the left"

File: MotionProfiling_Project/src/main/java/frc/robot/PathCreater.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    private Trajectory.Config config;
+    private ArrayList<Waypoint[]> paths;
+    private Trajectory[] trajectories;
+    private TankModifier modifier;
+
+    /**
+     * we configure the pathfinding and create new pathes, then we generate all the
+     * pathes we created
+     */
+    public PathCreater() {
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/PathCreater.java]: we create path acrording to index number, the first path is path 1...

File: MotionProfiling_Project/src/main/java/frc/robot/PathCreater.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+        }
+    }
+
+    /**
+     * we split the tank drive for bpth sides of the robot and get the trajectory of
+     * each one, first we get the right then the left trajectory
+     */
+    public Trajectory[] getTrajectories(int pathNumber) {
+        this.modifier = new TankModifier(trajectories[pathNumber]);
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/PathCreater.java]: getTrajectories to getSplitTrajectories

File: MotionProfiling_Project/src/main/java/frc/robot/PathCreater.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    /** writes all the paths to csv for quick generation of paths */
+    public void writeToCSV_AllPaths() {
+        for (int i = 0; i < this.paths.size(); i++) {
+            String s = "/home/lvuser/Path " + i + 1 + ".csv";
+            Pathfinder.writeToCSV(new File(s), this.trajectories[i]);
+        }
+    }
+}
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/PathCreater.java]: we need to add func for reading csv file to trajectory

File: MotionProfiling_Project/src/main/java/frc/robot/PathCreater.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        this.modifier.modify(RobotConstants.WHEEL_BASE_WIDTH);
+        Trajectory[] trajectories = { this.modifier.getRightTrajectory(), this.modifier.getLeftTrajectory() };
+        return trajectories;
+    }
+
+    /** writes all the paths to csv for quick generation of paths */
+    public void writeToCSV_AllPaths() {
+        for (int i = 0; i < this.paths.size(); i++) {
+            String s = "/home/lvuser/Path " + i + 1 + ".csv";
+            Pathfinder.writeToCSV(new File(s), this.trajectories[i]);
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/PathCreater.java]: "/home/lvuser/Path " to "/home/lvuser/Paths "

File: MotionProfiling_Project/src/main/java/frc/robot/PathCreater.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+    /**
+     * we split the tank drive for bpth sides of the robot and get the trajectory of
+     * each one, first we get the right then the left trajectory
+     */
+    public Trajectory[] getTrajectories(int pathNumber) {
+        this.modifier = new TankModifier(trajectories[pathNumber]);
+        this.modifier.modify(RobotConstants.WHEEL_BASE_WIDTH);
+        Trajectory[] trajectories = { this.modifier.getRightTrajectory(), this.modifier.getLeftTrajectory() };
+        return trajectories;
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/PathCreater.java]: trajectories to splitTrajectories

File: MotionProfiling_Project/src/main/java/frc/robot/FollowPath.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+ * this command uses the paths we generated on the path creater and uses it to
+ * perform the motion profiling
+ */
+public class FollowPath extends Command {
+
+  int pathNumber;
+  EncoderFollower right, left;
+  double leftCalculate, rightCalculate, gyroHeading, desiredHeading, angleDifference, turn, angleDiff;
+
+  /** we get the correct path and creat the encoder follower objects */
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/FollowPath.java]: the command gets path number and follows it

File: MotionProfiling_Project/src/main/java/frc/robot/FollowPath.java [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ * this command uses the paths we generated on the path creater and uses it to
+ * perform the motion profiling
+ */
+public class FollowPath extends Command {
+
+  int pathNumber;
+  EncoderFollower right, left;
+  double leftCalculate, rightCalculate, gyroHeading, desiredHeading, angleDifference, turn, angleDiff;
+
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/FollowPath.java]: private!!!

File: MotionProfiling_Project/src/main/java/frc/robot/FollowPath.java [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+    requires(Robot.driveTrain);
+    this.pathNumber = pathNumber;
+
+    this.left = new EncoderFollower(Robot.pathCreater.getTrajectories(this.pathNumber)[0]);
+    this.right = new EncoderFollower(Robot.pathCreater.getTrajectories(this.pathNumber)[1]);
+  }
+
+  @Override
+  /** we configure the encoder and the pidva to be ready of rmotion profiling */
+  protected void initialize() {
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/FollowPath.java]: /** we configure the encoder settings and the pidva values */

File: MotionProfiling_Project/src/main/java/frc/robot/FollowPath.java [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+        RobotConstants.WHEEL_DIAMETER);
+    this.left.configurePIDVA(RobotConstants.MotionProfilingPIDSettings.KP, 0,
+        RobotConstants.MotionProfilingPIDSettings.KD, 1 / RobotConstants.MotionProfilingPIDSettings.KV,
+        RobotConstants.MotionProfilingPIDSettings.KA);
+  }
+
+  @Override
+  /**
+   * we calculate the needed powere for the motion profiling, then we calculate
+   * the heading of the gyro to acount for the heading of the robot, the math is
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/FollowPath.java]: power, the math is the KP.  - תוריד

File: MotionProfiling_Project/src/main/java/frc/robot/FollowPath.java [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  @Override
+  protected boolean isFinished() {
+    return false;
+  }
+
+  @Override
+  protected void end() {
+  }
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/FollowPath.java]: motors are set to (0,0)

File: MotionProfiling_Project/src/main/java/frc/robot/FollowPath.java [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+
+  @Override
+  protected void interrupted() {
+  }
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/FollowPath.java]: interrupted() calls end

File: MotionProfiling_Project/src/main/java/frc/robot/FollowPath.java [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+
+    this.turn = RobotConstants.kP_TURN * (-1.0 / 80.0) * this.angleDifference;
+
+    Robot.driveTrain.tankDrive(this.leftCalculate + turn, this.rightCalculate - turn);
+  }
+
+  @Override
+  protected boolean isFinished() {
+    return false;
+  }
```
Comment (HilelAmiel) [inline MotionProfiling_Project/src/main/java/frc/robot/FollowPath.java]: if both of the encoders finished the tracking

**Review by HilelAmiel — COMMENTED** [review]
good work, there is more changes you need todo...

**Review by EliyaFishman — APPROVED** [review]

**Pull request merged**